### PR TITLE
DM-37933: Refactor the prepuller

### PR DESCRIPTION
**Pull request by rra**
Extract the prepuller logic for determining remote and cached images and calculating the prepull targets, menus, and prepull status into a new ImageService class. Rewrite it on top of the new RSPImageTag and RSPImage abstract data types and their collections, which can now do most of the filtering work. The new image service is now used in handlers where previously the PrepullerAribtrator was used.

This new image service only asks for digests for images that are prepull targets, so has less impact on the Docker registry API. However, it doesn't lock in that decision; we can decide to retrieve hashes for all images with some minor additional work. The public API for available images drops the human-readable descriptions for alias tags (I think they add more clutter than useful information) and distinguishes between the main tag and the aliases using the new RSPImage data structure.

Ensuring that every pod will be launched with a reference with a digest will be added in a subsequent commit.

The image service now uses a single delay for polling both Docker and Kubernetes (which now may be a bit long, but it can be updated in the future with experience).

What remains of the prepuller is only the logic to ask the image service for images to prepull and start tasks to do so. Simplify its coordination with the image service by using asyncio.Event instead of polling and timeouts, so the prepuller is kicked off each time the image status has been refreshed. The prepuller creates one background task per node with missing images and pulls those images in the same sort order that is used in the menu (with the minor variation that all alias tags are pulled first in alphabetical order rather than forcing the recommended tag first).

The resulting prepuller is small enough that it can be collapsed into a single class.

### CODE REVIEW

File: src/jupyterlabcontroller/assets/form_template.txt [hunk 1]
```diff
@@ -39,7 +39,7 @@ function selectDropdown() {
     </label><br />
     <select name="image_dropdown" onclick="selectDropdown()">
 {% for i in all_images -%}
-        <option value="{{ i.path }}">{{ i.name }}</option>
+        <option value="{{ i.reference }}">{{ i.name }}</option>
```
Comment (athornton) [inline src/jupyterlabcontroller/assets/form_template.txt:42]: Much better.

File: src/jupyterlabcontroller/models/domain/kubernetes.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import re
+from dataclasses import dataclass
+from typing import Self
+
+from kubernetes_asyncio.client import V1ContainerImage
+
+__all__ = ["KubernetesNodeImage"]
+
+# Regex fragments used for Docker reference parsing.
```
Comment (athornton) [inline src/jupyterlabcontroller/models/domain/kubernetes.py:13]: Glad Someone Who Isn't Me rewrote this into a set of regexes.

File: src/jupyterlabcontroller/models/v1/prepuller.py [hunk 3]
```diff
@@ -1,152 +1,104 @@
...
+class PrepullerImageStatus(BaseModel):
+    prepulled: list[NodeImage] = Field([], title="Successfully cached images")
+    pending: list[NodeImage] = Field(
+        [],
+        title="Images not yet cached on all eligible nodes",
+    )
 
 
-class Node(CamelCaseModel):
+class Node(BaseModel):
```
Comment (athornton) [inline src/jupyterlabcontroller/models/v1/prepuller.py:78]: Don't forget to update sqr-066 to reflect any changes made here to the externally-visible models.

File: src/jupyterlabcontroller/services/prepuller.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        # easier for the test suite.)
+        await self._image_service.prepuller_wait()
+        await self._scheduler.spawn(self._prepull_loop())
+        self._running = True
+
+    async def stop(self) -> None:
+        if not self._running:
+            self._logger.info("Prepuller tasks were already stopped")
+        self._logger.info("Cancelling prepuller tasks")
+        await self._scheduler.close()
```
Comment (athornton) [inline src/jupyterlabcontroller/services/prepuller.py:76]: Is closing a closed scheduler OK, or do you need to check the scheduler's status (or just return) if not self._running ?

File: src/jupyterlabcontroller/services/prepuller.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                    command=["/bin/sleep", "5"],
+                    image=image.reference_with_digest,
+                    working_dir="/tmp",
+                )
+            ],
+            node_name=node,
+            restart_policy="Never",
+        )
+
+    async def _prepull_image(self, image: RSPImage, node: str) -> None:
```
Comment (athornton) [inline src/jupyterlabcontroller/services/prepuller.py:126]: Maybe put the last 8 chars of the nodename in the pod name too?  It's nice to be able to see without having to dig through the YAML which node a misbehaving prepull pod should be on.

**Review by athornton — APPROVED** [review]
A couple of questions and suggestions, but basically seems fine.

File: src/jupyterlabcontroller/models/v1/prepuller.py (hunk 3 above)
Comment (rra) [reply src/jupyterlabcontroller/models/v1/prepuller.py:78]: Yeah, good call, draft PR here: https://github.com/lsst-sqre/sqr-066/pull/3

This currently has some stuff in it from upcoming PRs.

File: src/jupyterlabcontroller/services/prepuller.py (hunk 4 above)
Comment (rra) [reply src/jupyterlabcontroller/services/prepuller.py:76]: Good catch, it looks like you can't reuse a scheduler. Will fix that.

File: src/jupyterlabcontroller/services/prepuller.py (hunk 5 above)
Comment (rra) [reply src/jupyterlabcontroller/services/prepuller.py:128]: Ack, sorry, we talked about this and I forgot that I committed something I was playing around with and didn't go back and fix it. Should now be fixed properly based on the Kubernetes object naming rules.

**Pull request merged**